File: javascript/specs/elements/field_with_attributes.spec.js
import { parse } from '../../lib/esm/main.js';
import { Attribute } from '../../lib/esm/elements/attribute.js';
import { MissingAttribute } from '../../lib/esm/elements/missing/missing_attribute.js';

describe('Field', () => {
    let field;
    
    beforeEach(() => {
        field =  parse(`
field:
attribute = value
other = value
            `.trim()).field('field');
        });
        
        it('is untouched after initialization', () => {
            expect(field._touched).toBeUndefined();
        });
        
        it('has only untouched attributes after initialization', () => {
            for (let attribute of field.attributes()) {
                expect(attribute._touched).toBeUndefined();
            }
        });
        
        it('has allAttributesRequired disabled by default', () => {
            expect(field._allAttributesRequired).toBe(false);
        });
        
        describe('attribute()', () => {
            describe('fetching an existing element', () => {
                let attribute;
                
                beforeEach(() => {
                    attribute = field.attribute('attribute');
                });
                
                it('returns a Attribute', () => {
                    expect(attribute).toBeInstanceOf(Attribute);
                });
                
                it('returns the right attribute', () => {
                    expect(attribute.stringKey()).toEqual('attribute');
                });
            });
            
            describe('fetching a missing element', () => {
                it('returns a MissingAttribute', () => {
                    const missingAttribute = field.attribute('missing');
                    expect(missingAttribute).toBeInstanceOf(MissingAttribute);
                });
            });
        });
        
        describe('allAttributesRequired()', () => {
            it('sets the _allAttributesRequired property to true', () => {
                field.allAttributesRequired();
                expect(field._allAttributesRequired).toBe(true);
            });
            
            describe('passing true explicitly', () => {
                it('sets the _allAttributesRequired property to true', () => {
                    field.allAttributesRequired(true);
                    expect(field._allAttributesRequired).toBe(true);
                });
            });
            
            describe('passing false explicitly', () => {
                it('sets the _allAttributesRequired property back to false', () => {
                    field.allAttributesRequired(true);
                    field.allAttributesRequired(false);
                    expect(field._allAttributesRequired).toBe(false);
                });
            });
        });
        
        describe('toString()', () => {
            it('returns a debug abstraction', () => {
                expect(field.toString()).toEqual('[object Field key=field attributes=2]');
            });
        });
        
        describe('toStringTag symbol', () => {
            it('returns a custom tag', () => {
                expect(Object.prototype.toString.call(field)).toEqual('[object Field]');
            });
        });
        
        describe('touch()', () => {
            beforeEach(() => {
                field.touch();
            });
            
            it('touches the field', () => {
                expect(field._touched).toBe(true);
            });
            
            it('touches the attributes', () => {
                for (const attribute of field.attributes()) {
                    expect(attribute._touched).toBe(true);
                }
            });
        });
    });

File: ruby/spec/elements/field_with_attributes.spec.rb
# frozen_string_literal: true

input = <<~DOC.strip
field:
attribute = value
other = value
DOC

describe Enolib::Field do
  before(:each) do
    @field = Enolib.parse(input).field
  end

  it 'is untouched after initialization' do
    expect(@field.instance_variable_defined?(:@touched)).to be false
  end

  it 'has only untouched attributes after initialization' do
    @field.attributes.each do |attribute|
      expect(attribute.instance_variable_defined?(:@touched)).to be false
    end
  end

  it 'has all_attributes_required disabled by default' do
    expect(@field.instance_variable_get(:@all_attributes_required)).to be false
  end

  describe '#attribute' do
    it 'returns an attribute' do
      expect(@field.attribute('attribute')).to be_a(Enolib::Attribute)
    end

    it 'returns the right attribute' do
      expect(@field.attribute('attribute').string_key).to eq('attribute')
    end
  end

  describe '#attributes' do
    before(:each) do
      @attributes = @field.attributes
    end

    it 'returns all attributes' do
      expect(@attributes.length).to be 2
    end

    it 'returns attributes' do
      @attributes.each do |attribute|
        expect(attribute).to be_a(Enolib::Attribute)
      end
    end

    it 'touches the field' do
      expect(@field.instance_variable_defined?(:@touched)).to be true
    end

    it 'does not touch the attributes' do
      @attributes.each do |attribute|
        expect(attribute.instance_variable_defined?(:@touched)).to be false
      end
    end
  end

  describe '#attribute' do
    context 'when the attribute does not exist' do
      it 'returns a missing proxy instance' do
        expect(@field.attribute('missing')).to be_a(Enolib::MissingAttribute)
      end
    end
  end

  describe '#touch' do
    before(:each) do
      @field.touch
    end

    it 'touches the field' do
      expect(@field.instance_variable_defined?(:@touched)).to be true
    end

    it 'touches the field attributes' do
      @field.attributes.each do |attribute|
        expect(attribute.instance_variable_defined?(:@touched)).to be true
      end
    end
  end

  describe '#to_s' do
    it 'returns a debug abstraction' do
      expect(@field.to_s).to eq('#<Enolib::Field key=field attributes=2>')
    end
  end
end
